**Антон, привет!**

Меня зовут Соколова Александра, я буду проверять твой проект. Приятно познакомиться:)
    
Предлагаю общаться на «ты», но если тебе не очень комфортно - дай, пожалуйста, знать. 

Ты уже проделал большую работу над проектом, давай сделаем его еще лучше. Ниже ты найдешь мои комментарии, пожалуйста, не перемещай, не изменяй и не удаляй их. Давай использовать цветовую разметку:

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> Так выделены самые важные замечания, их нужно обязательно проработать. </div>

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> Так выделены небольшие замечания, некритичные ошибки или развивающие рекомендации на будущее. 

</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Все отлично:) </div>

Давай работать над проектом в диалоге: если ты что-то меняешь или отвечаешь на мои комменатри — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера №2:</b> 
    
Антон, привет!
    
Новые комментарии будут обозначены, как: **Комментарий  ревьюера №2**

</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера №3:</b> 
    
Антон, привет!
    
Новые комментарии будут обозначены, как: **Комментарий  ревьюера №3** =)

</div>

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> 
    
На этом этапе было бы здорово немного расширить оформление проекта - написать его своими словами. 
 
***
    
Оформление принято делать к каждому проекту, обычно прописывают название, цели и задачи исследования. Можно добавить этапы работы, а также итоги исследования (что было достигнуто в процессе, к каким выводам пришли, что нашли в данных).
   
</div>

- Структура данных:
сalls — количество звонков,
minutes — суммарная длительность звонков в минутах,
messages — количество sms-сообщений,
mb_used — израсходованный интернет-трафик в Мб,
is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

- План работы:
Изучить файл с данными.
Разбить данные на три выборки: обучающая, валидационная и тестовая.
Исследовать три модели классификации: Решающее дерево, Случайный лес и Логистическая регрессия.
Найти оптимальные параметры для каждой модели и выбрать одну из них для обучения модели.
Оценить точность обученной модели.
Оценить адекватность модели.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера №2:</b> 
    
Когда будешь готовить портфолио, обычно его размещают на Github, хорошее введение - это важная часть. 
    
На плюсовых программах Практикума есть Карьерный трек, там hr-специалисты учат готовить портфолио и дают рекомендации в отношении оформления. Для каждого проекта советуют подготовить описание (обычно в README.md сохраняют). Т.е. это базовая составляющая хорошего портфолио 

</div>

## Откройте и изучите файл

In [1]:
import pandas as pd
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier

from sklearn.ensemble import RandomForestClassifier

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

import warnings
warnings.filterwarnings('ignore') 

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> 
    
Ты загружаешь в проект библиотеки, которые не используешь, например:
    
- DecisionTreeRegressor
- RandomForestRegressor

Их лучше удалить. Они не требуются потому что в проекте стоит задача классификации. 
</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера №2:</b> 
    
👍

</div>

In [2]:
users_behavior = pd.read_csv('/datasets/users_behavior.csv')

In [3]:
users_behavior.shape

(3214, 5)

In [4]:
users_behavior.head(10)

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0
5   58.0   344.56      21.0  15823.37         0
6   57.0   431.64      20.0   3738.90         1
7   15.0   132.40       6.0  21911.60         0
8    7.0    43.39       3.0   2538.67         1
9   90.0   665.41      38.0  17358.61         0

In [5]:
users_behavior.info

<bound method DataFrame.info of       calls  minutes  messages   mb_used  is_ultra
0      40.0   311.90      83.0  19915.42         0
1      85.0   516.75      56.0  22696.96         0
2      77.0   467.66      86.0  21060.45         0
3     106.0   745.53      81.0   8437.39         1
4      66.0   418.74       1.0  14502.75         0
...     ...      ...       ...       ...       ...
3209  122.0   910.98      20.0  35124.90         1
3210   25.0   190.36       0.0   3275.61         0
3211   97.0   634.44      70.0  13974.06         0
3212   64.0   462.32      90.0  31239.78         0
3213   80.0   566.09       6.0  29480.52         1

[3214 rows x 5 columns]>

- Перед нами стоит задача классификации. Датасет состоит из 3 214 объектов и 5 признаков. За целевой признак берем столбец is_ultra, т.е. модель, которую мы попытаемся построить, будет предсказывать значение 1 если клиенту нужно предложить тариф "Ультра" или 0 если тариф "Смарт". 

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b>  

Удачное начало проекта!👍
    
Изучать данные - очень хорошая практика, ниже в комментарии оставляю список того минимума, который стоит смотреть перед началом работы с данными (часть ты уже отразил в проекте выше 👍).
    

*** 
- Использовать `describe()` для выведения информации и поиска инсайтов в данных
- Критично посмотреть на данные с точки зрения наличия пропусков в признаках
- Явно обозначить какую задачу будем решать: с учителем / без учителя, классификации / регрессии и тд
- Проверить соотношение классов. В моделях бинарной классификации имеет смысл смотреть на баланс классов. Два взгляда (с разных сторон) на эту проблему тут: https://alexanderdyakonov.wordpress.com/2021/05/27/imbalance/ и https://habr.com/ru/post/349078/
- Построить и визуализировать матрицу корреляции признаков. При высокой их степени скоррелированности возникает проблема  мультиколлиниарности признаков. 
- Написать вывод
    
PS - Это несколько моментов "на будущее" :)
    
</div>

## Разбейте данные на выборки

Спрятанной тестовой выборки в задании нет, поэтому стоит разбить исходный набор данных на три части: обучающую, валидационную и тестовую (60/20/20).

In [6]:
df_train, df_valid = train_test_split(users_behavior, test_size=0.4, random_state=12345)
df_valid, df_test = train_test_split(df_valid, test_size=0.5, random_state=12345)

In [7]:
features_train = df_train.drop(['is_ultra'], axis=1)
target_train = df_train['is_ultra']

features_valid = df_valid.drop(['is_ultra'], axis=1)
target_valid = df_valid['is_ultra']

features_test = df_test.drop(['is_ultra'], axis=1)
target_test = df_test['is_ultra']

<div class="alert alert-success">
<b>Комментарий ревьюера: </b>

Хорошо, указан параметр `random_state`👍 Воспроизводимость результатов разбиения на обучающую (тренировочную) / тестовую / валидационную выборок обеспечена, значит при всех последующих запусках кода подвыборки будут идентичными
    
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера:</b> 

После разбиения лучше вывести количество строк, получившиеся в выборках. Рекомендую использовать `.shape`
    
</div>

In [8]:
print (features_train.shape)
print (features_valid.shape)
print (features_test.shape)

(1928, 4)
(643, 4)
(643, 4)


<div class="alert alert-success">
<b>✔️ Комментарий ревьюера №2:</b> 
    
👍
    
Теперь можно проверить корректность деления на выборки.

</div>

<div class="alert alert-info">
<b>Комментарий ревьюера:</b>

В функции `train_test_split()` можно использовать параметр `stratify = y` , где y – соответствующая целевая переменная. Данные разделятся на два кластера (наблюдения класса 0 и наблюдения класса 1), и в валидационную выборку попадут по 20% наблюдений из каждого кластера:
    
То есть, если соотношение классов в оригинальной выборке было 2:5, то в обучающей и валидационной выборках оно сохранится таким же. Такой подход в чем-то решает проблему несбалансированных классов.

Функция train_test_split( ) строена так, что мы не сталкиваемся с подобной проблемой. Но я все же пишу про этот параметр, чтобы обратить внимание на проблему дисбаланса :)

    
</div>

## Исследуйте модели

Исследуем возможные модели. 

Начнем с Дерева. Изменим в цикле гиперпараметр max_depth.

In [9]:
best_depth = 0
best_accuracy = 0

for depth in range(1, 21, 2):
    model = DecisionTreeClassifier(random_state=12345, max_depth = depth)
    model.fit(features_train,target_train) 
    predictions_valid = model.predict(features_valid) 
    accuracy = accuracy_score(target_valid, predictions_valid)
    
    print('Глубина дерева=', depth,'Точность',accuracy)
    if accuracy > best_accuracy:
        best_depth = depth
        best_accuracy = accuracy

print('Лучшая глубина дерева =', best_depth,'Лучшая точность', best_accuracy) 

Глубина дерева= 1 Точность 0.7542768273716952
Глубина дерева= 3 Точность 0.7853810264385692
Глубина дерева= 5 Точность 0.7791601866251944
Глубина дерева= 7 Точность 0.7822706065318819
Глубина дерева= 9 Точность 0.7822706065318819
Глубина дерева= 11 Точность 0.7620528771384136
Глубина дерева= 13 Точность 0.7558320373250389
Глубина дерева= 15 Точность 0.7465007776049767
Глубина дерева= 17 Точность 0.7356143079315708
Глубина дерева= 19 Точность 0.7278382581648523
Лучшая глубина дерева = 3 Лучшая точность 0.7853810264385692


- Лучшая глубина дерева = 3. Лучшая точность 0.7853810264385692.

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера:</b> 

В разные переменные записано одно:
    
    result = accuracy_score(target_valid, predictions_valid) 
    accuracy = accuracy_score(target_valid, predictions_valid)
    
    
    
</div>

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера №2:</b> 
    
К сожалению, появляется ошибка:
    
    NameError: name 'result' is not defined

Боюсь, ты удалил переменную, которую дальше используешь в print().
    
***    
    
Было так:
    
    result = accuracy_score(target_valid, predictions_valid) - эту строку ты удалил
    accuracy = accuracy_score(target_valid, predictions_valid)
    
Но дальше используешь `result`, т.е. тут нужно поправить:
    
    print('Глубина дерева =', depth, 'Точность', result)

</div>


<div class="alert alert-success">
<b>✔️ Заключительный комментарий ревьюера №2:</b> 
    
Антон, спасибо, что внес правки!
    
В т.к. выходит ошибка в коде, пока вынуждена вернуть проект на доработку =(
    
</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера №3:</b> 
    
Теперь код загружается успешно
</div>

Случайный лес. Изменим в цикле гиперпараметр n_estimators.

In [10]:
best_estim = 0
best_accuracy = 0

for estim in range(1,21, 2):
    model = RandomForestClassifier(random_state=12345, n_estimators=estim)
    model.fit(features_train, target_train)
    valid_predictions = model.predict(features_valid)
    accuracy = accuracy_score(target_valid, predictions_valid)
    print('Количество деревьев =', estim,'Точность',accuracy)
    if accuracy > best_accuracy:
        best_estim = estim
        best_accuracy = accuracy

print('Лучшее количество деревьев =', best_estim,'Лучшая точность', best_accuracy)

Количество деревьев = 1 Точность 0.7278382581648523
Количество деревьев = 3 Точность 0.7278382581648523
Количество деревьев = 5 Точность 0.7278382581648523
Количество деревьев = 7 Точность 0.7278382581648523
Количество деревьев = 9 Точность 0.7278382581648523
Количество деревьев = 11 Точность 0.7278382581648523
Количество деревьев = 13 Точность 0.7278382581648523
Количество деревьев = 15 Точность 0.7278382581648523
Количество деревьев = 17 Точность 0.7278382581648523
Количество деревьев = 19 Точность 0.7278382581648523
Лучшее количество деревьев = 1 Лучшая точность 0.7278382581648523


- При различном количестве деревьев точность одинакова: 0.7278382581648523.

Логистическая регрессия. Изменим в цикле гиперпараметр max_iter.

In [11]:
best_max_iter = 0
best_accuracy = 0

for max_iter in range(1, 21, 2):
    model = LogisticRegression(random_state=12345, max_iter=max_iter)
    model.fit(features_train, target_train)
    valid_predictions = model.predict(features_valid)
    accuracy = accuracy_score(target_valid, valid_predictions)
    print('Количество итерраций =', max_iter,'Точность',accuracy)
    if accuracy > best_accuracy:
        best_max_iter = max_iter
        best_accuracy = accuracy
print('Лучшее количество итераций =', best_max_iter,'Лучшая точность', best_accuracy)        

Количество итерраций = 1 Точность 0.7060653188180405
Количество итерраций = 3 Точность 0.7060653188180405
Количество итерраций = 5 Точность 0.7060653188180405
Количество итерраций = 7 Точность 0.7060653188180405
Количество итерраций = 9 Точность 0.7060653188180405
Количество итерраций = 11 Точность 0.7060653188180405
Количество итерраций = 13 Точность 0.7060653188180405
Количество итерраций = 15 Точность 0.7060653188180405
Количество итерраций = 17 Точность 0.7060653188180405
Количество итерраций = 19 Точность 0.7076205287713841
Лучшее количество итераций = 19 Лучшая точность 0.7076205287713841


- Лучшее количество итераций = 19. Лучшая точность 0.7076205287713841.

Попробуем улучшить результат изменив гиперпараметр solver – алгоритм задачи оптимизации.

In [12]:
best_solver = ''
best_accuracy = 0

for solve in {'newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga'}:
    model = LogisticRegression(random_state=12345, solver=solve)
    model.fit(features_train, target_train)
    valid_predictions = model.predict(features_valid)
    accuracy = accuracy_score(target_valid, valid_predictions)
    print('Solver-', solve,'. ''Точность-',accuracy)
    if accuracy > best_accuracy:
        best_solver= solve
        best_accuracy = accuracy

print('Лучший solver-', best_solver,'. ''Лучшая точность-', best_accuracy) 

Solver- lbfgs . Точность- 0.7107309486780715
Solver- saga . Точность- 0.7060653188180405
Solver- newton-cg . Точность- 0.7558320373250389
Solver- liblinear . Точность- 0.7091757387247278
Solver- sag . Точность- 0.7060653188180405
Лучший solver- newton-cg . Лучшая точность- 0.7558320373250389


- Лучшим solver оказался newton-cg а не lbfgs, который стоит по-умолчанию.

Вывод:
Точность модели решающего дерева при глубине 3 равна 0.785. 
Модель случайного леса дает один результат- 0.728. 
Точность модели логистической регресии равна 0.708. 

Для решения задачи оптимизации, модель логистической регрессии гораздо менее точна, чем дерево решений и случайный лес.

Так как выборка небольшая, то лучший результат показывает решающее древо с гиперпараметром max_depth=3.

<div class="alert alert-success">
<b>Комментарий ревьюера: </b>
    
Молодец, что рассматриваешь разные модели. 🔥
    
И добавил развернутый вывод к этапу.    
    
</div>

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> 
    
На этом шаге **в рамках одной модели** можно перебрать хотя бы два гиперпараметра.
    
**Т.е. задача состоит в том, чтобы хотя бы в одной из моделей попробовать указать несколько гиперпараметров (два или больше).** Это делают, чтобы улучшить работу модели и, как результат, метрику.  
    
***
    
Чтобы узнать, какие гиперпараметры можно применить к модели, стоит заглянуть в документацию, например: 
    
`RandomForestClassifier:`    
https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn-ensemble-randomforestclassifier     
    
`DecisionTreeClassifier:`
https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html#sklearn-tree-decisiontreeclassifier
 
В документации хорошо прописаны дефолтные значения гиперпараметров, их тоже можо настроить.     


</div>

<div class="alert alert-info"> <b>Комментарий ревьюера:</b> 
    
**Полезные материалы (на будущее):**
    
Для подбора оптимальных параметров можно использовать функции Gridsearchcv или Randomizedsearchcv. 
    
1. Статья про разницу между ними: 
    https://towardsdatascience.com/gridsearchcv-or-randomsearchcv-5aa4acf5348c#:~:text=RandomSearchCV%20has%20the%20same%20purpose,parameters%20to%20improve%20your%20model.&text=The%20main%20difference%20between%20the,want%20to%20sample%20and%20test. 
    
2. И просто очень большая статья по их применению: 
    https://machinelearningmastery.com/hyperparameter-optimization-with-random-search-and-grid-search/ 
    
3. Совсем обьемная статья с огромным количеством библиотек для подбора гиперпараметров:
    https://neptune.ai/blog/hyperparameter-tuning-in-python-complete-guide/. 
    
4. Ускоренная версия GridsearchCV 
    https://towardsdatascience.com/20x-times-faster-grid-search-cross-validation-19ef01409b7c.    
    

## Проверьте модель на тестовой выборке

In [13]:
features_test = df_test.drop(['is_ultra'], axis=1)
target_test = df_test['is_ultra']

model = DecisionTreeClassifier(random_state=12345, max_depth = 3)
model.fit(features_train, target_train)

predictions_test = model.predict(features_test)
result = accuracy_score(target_test, predictions_test)

print('Accuracy DecisionTreeClassifier= ', result)

Accuracy DecisionTreeClassifier=  0.7791601866251944


- На тестовой выборке наша модель решающего дерева показывает приближенное к тренировочной выборке значение accuracy. Таким образом, мы не видим здесь явной проблемы с переобучением, и можем считать, что нам удалось построить модель для задачи бинарной классификации, которая выберет подходящий тариф.



<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b>

Хороший результат 👍

</div>

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера:</b> 

Когда добавишь второй гиперпараметр в этапе выше, тут тоже стоит его отразить. Метрика должна улучшиться.
    
 
</div>

## (бонус) Проверьте модели на адекватность

Для проверки используем модель DummyClassifier. DummyClassifier делает прогнозы, которые игнорируют входные функции.
Этот классификатор служит простой основой для сравнения с другими, более сложными классификаторами.

In [14]:
from sklearn.dummy import DummyClassifier
model_dc = DummyClassifier(strategy='most_frequent', random_state=12345)
model_dc.fit(features_train, target_train)
result_dc = model_dc.score(features_valid, target_valid)
print('Accuracy DummyClassifier:', result_dc)

Accuracy DummyClassifier: 0.7060653188180405


- Accuracy модели DummyClassifier ниже, чем у модели DecisionTreeClassifier, из этого следует что наша модель адекватна и эффективна.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>

Очень здорово, что ты уже владеешь инструментами `DummyClassifier`  🔥
 
***    
    
p.s. (на перспективу!): можно было бы построить `Confusion Matrix`, чтобы детально посмотреть, где ошибается модель. Подробнее о Confusion Matrix: https://neptune.ai/blog/evaluation-metrics-binary-classification        
    
</div>

***

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера❌:</b> 
    
Давай сформируем итоговый вывод про проекту 😊  
    
В выводе можно отметить следующие пункты:
- какие у исследования цели и задачи     
- какие модели рассматривались
- какая модель показала лучшие результаты (гиперпараметры, метрика)
- результаты проверки лучшей модели на тестовой выборке   
- любые другие инсайты, которые ты заметил в ходе исследования
    
</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера №2:</b> 
    
Хорошее, что есть резюме проекта👍

</div>

***

<div class="alert alert-success">
<b>✔️ Заключительный комментарий ревьюера:</b> 
    
Антон, спасибо за проделанную работу! 
    
Молодец, что начал делать проект и рассмотрел разные модели, движение в позитивную сторону положено👍 
    
Здорово, что изучил материалы сверх теории спринта и знаешь, как проверить модель на адекватность с помощью DummyClassifier.🔥  
      
***    
 
В работе есть несколько моментов, которые лучше доработать:  
    
1. **Этап исследования моделей** - в соответствии с красным комментарием
2. Сформировать итоговый вывод
  
***

Возвращаю тебе работу, чтобы ты смог доделать проект и задать вопросы)
    
Успехов!

</div>

<div class="alert alert-success">
<b>✔️ Заключительный комментарий ревьюера №2:</b> 
    
Антон, у тебя получилось хорошее исследование.
  
Принимаю твою работу и желаю удачи в учебе и новых проектах 😊

</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75
